Feature/issue 19

dzoccarato opened the pull request:
- [x] Related GitHub issue(s) linked in PR description
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [ ] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

Issue; #19 

**Notes:**
1. This PR depends on https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/60
2. There is an issue with test cases because of `AndcultureCode.CSharp.Core.Models.Errors` namespace changes
3. As for PR 60 on AndcultureCode.CSharp.Core. (https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/60) , `FindAll` has an overload to accept group by clause.
Made this choice because `GroupBy` returns IQueryable<IGrouping<TKey, TSource>>` instead of `IQueryable<TSource>` (https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs)

Is this acceptable for you?

dzoccarato commented:
Build fails because of error "No overload for method 'FindAll' takes 8 arguments", but this is caused by https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/60 that isn't merged, so the AndcultureCode.CSharp.Core package not contains this overload

wintondeshong commented on src/AndcultureCode.CSharp.Conductors/RepositoryConductor.cs:
```diff
@@ -310,6 +310,31 @@ public class RepositoryConductor<T> : Conductor, IRepositoryConductor<T>
+        /// <param name="ignoreQueryFilters"></param>
+        /// <param name="asNoTracking">Ignore change tracking on the result. Set <code>true</code> for read-only operations.</param>
+        /// <returns></returns>
+        public IResult<IQueryable<IGrouping<TKey, T>>> FindAll<TKey>(
```
@dzoccarato please accompany this new method with some tests. See 'Tests/RepositoryConductorTest`

dzoccarato replied:
should done in https://github.com/AndcultureCode/AndcultureCode.CSharp.Conductors/pull/24/commits/e4f05338acabcce9e83354b1a9b3ac0476924ca1

wintondeshong commented on src/AndcultureCode.CSharp.Conductors/RepositoryReadConductor.cs:
```diff
@@ -71,6 +71,30 @@ public RepositoryReadConductor(IRepository<T> repository)
+        /// <param name="ignoreQueryFilters">If true, global query filters will be ignored for this query.</param>
+        /// <param name="asNoTracking">Ignore change tracking on the result. Set <code>true</code> for read-only operations.</param>
+        /// <returns></returns>
+        public IResult<IQueryable<IGrouping<TKey, T>>> FindAll<TKey>(
```
@dzoccarato please accompany this new method with some tests.

dzoccarato replied:
should done in https://github.com/AndcultureCode/AndcultureCode.CSharp.Conductors/pull/24/commits/e4f05338acabcce9e83354b1a9b3ac0476924ca1

wintondeshong reviewed (changes requested):
@dzoccarato thank you for this contribution! if you wouldn't mind, could you add some tests before we merge it?

dzoccarato commented:
@wintondeshong , thanks for your feedback.
Can i make you a question? In order to add some tests, should I do some changes also on this repo: https://github.com/AndcultureCode/AndcultureCode.CSharp.Testing ?
Thanks!

dzoccarato commented:
@wintondeshong I've added the implementation of https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/60/commits/dad1bdf36b712135174b75feb64ee006af933094 and added test cases for it, would you like to take a look? thanks!

dzoccarato commented:
Hi @wintondeshong I've update the pakages after merges of https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/60
Would you like to take a look ?

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+        }
+
+        [Fact]
+        public void FindAll_GroupBy_When_FindAll_HasErrors_Then_Returns_An_Errors()
```
Typo: `Then_Returns_An_Errors` would read better as just `Then_Returns_Errors`

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+        public void FindAll_GroupBy_When_FindAll_HasErrors_Then_Returns_An_Errors()
+        {
+            // Arrange
+            var entities = new List<Entity>();
```
Is this used anywhere in the test?If not, please remove the unnecessary setup

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+            );
+
+            // Assert
+            result.ShouldHaveErrorsFor(BASIC_ERROR_KEY);
```
Minor: I'm sure you referenced an existing test for this, but I believe we have a `ShouldHaveBasicError()` result matcher now.

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+        public void FindAll_GroupBy_When_FindAll_Not_Find_Anyting_And_Succeeds_Then_Returns_An_Empty_IQueryable()
+        {
+            // Arrange
+            var entities = new List<Entity>();
```
Same here - is this used? If not, please remove it.

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+                    It.IsAny<int?>(), // take
+                    It.IsAny<bool?>(), // ignoreQueryFilters
+                    It.IsAny<bool>() // asNoTracking
+            )).Returns(new Result<IQueryable<IGrouping<long, Entity>>>());
```
There's a `ReturnsGivenResult` extension method which should wrap the instantiation of the `Result` object for you - you'll still need to give it the queryable, somehow (usually we use an empty `List.AsQueryable()`, but the `IGrouping` type might right slightly different setup)

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+                    new FindAllMokedEntity { Id = 0, },
+                    new FindAllMokedEntity { Id = 1, },
+                    new FindAllMokedEntity { Id = 2, },
+                    new FindAllMokedEntity { Id = 3, }
```
It would be awesome to leverage factories here! Take a look at our `ConnectionFactory` for an example. There is a `BuildList<T>` method in our `BaseUnitTest` class that instantiates multiple models that have a registered factory, which should help clean up some of this setup.

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+        }
+
+        [Fact]
+        public void FindAll_GroupBy_With_Selector_When_FindAll_Not_Find_Anyting_And_Succeeds_Then_Returns_An_Empty_IQueryable()
```
Instead of saying "Not_Find_Anything", this test name might be more succinctly written:

`FindAll_GroupBy_Given_Selector_When_No_Records_Found_Returns_Empty_List`

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+                groupBy: o => o.Id,
+                groupBySelector: (o, k) => new GroupingResult { Key = o, Value = k.Count() },
+                includeProperties: null,
+                skip: null,
```
For tests that don't rely on these values, it would probably help improve readability and reduce confusion for a reader if only the relevant parameters were being sent into the system under test 👍 I believe all of these have a default value, so nothing should be required.

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+            result.ResultObject.ShouldBeNull();
+        }
+
+
```
Nitpick: Extra new line

brandongregoryscott commented on test/AndcultureCode.CSharp.Conductors.Tests/RepositoryConductorTest.cs:
```diff
@@ -556,5 +566,273 @@ public void Delete_When_Delete_Succeeds_Then_Returns_True()
+        /// <summary>
+        /// moke class used in groupBySelector of tests below
+        /// </summary>
+        internal class GroupingResult
```
Since this class is used for this set of tests but isn't necessarily a model that could be reused in tests, I would be inclined to wrap this in a `Setup` region nested right at the start of the `FindAll` region so it's easy to locate for a reader but separated enough from the actual test methods.

brandongregoryscott reviewed (changes requested):
Thanks for the PR, @dzoccarato! I left a number of comments for typos, suggested factors & cleanup. Let me know if you need any additional direction or examples 👍

dzoccarato commented:
> Thanks for the PR, @dzoccarato! I left a number of comments for typos, suggested factors & cleanup. Let me know if you need any additional direction or examples 👍

Hi @brandongregoryscott ! Thanks for te review! I've update this PR with according your comments.
There is a change that I've made where I'd like to have your feedback.

I'm talking about changes you suggest on: https://github.com/AndcultureCode/AndcultureCode.CSharp.Conductors/pull/24#discussion_r511233584.
In order to be able to use `BuildList<>` factory with `TestEntity` is necessary to register a factory with `FactoryExtensions.DefineFactory`, otherwise `BuildList<>` throw an exception.

I've take a look how `BuildList<>` (https://github.com/AndcultureCode/AndcultureCode.CSharp.Testing/blob/589d510ab1520805ba76a20162b5c0963abd3f19/src/AndcultureCode.CSharp.Testing/Tests/BaseTest.cs) works and, under the hood, there is a cycle that call `FactoryExtensions.Build<T>() => BuildObject<T>(new List<Action<T>>(), typeof(T).Name)`, that perform this type of check:
```cs
// https://github.com/AndcultureCode/AndcultureCode.CSharp.Testing/blob/589d510ab1520805ba76a20162b5c0963abd3f19/src/AndcultureCode.CSharp.Testing/Factories/FactoryExtensions.cs#L78-L91

if (!namedBuilders.ContainsKey(keyName))
{
    throw new MissingFactoryException(name + " is not yet registered as a factory (Tests\\Testing\\Factories).  You can only build objects for factories that have been created");
}
```
`FactoryExtensions.DefineFactory<>(...)` requires a `Factory` to be able to register a factory on `namedBuilders`.

For this reason, I've create a new class `TestEntityFactory` under new Factory folder, otherwise I'm not able to use `BuildList<>` facade.
But in my opinion this isn't the right project to do that: correct me if I'm wrong, but this Factory class should move to `AndcultureCode.CSharp.Testing/src/AndcultureCode.CSharp.Testing/Factories` project.

I've decided to commit it otherwise this project can't build, but If you prefer I can fork `AndcultureCode.CSharp.Testing`, move this Factory on it and make a new PR.

brandongregoryscott commented:
Thanks for taking a dive into our other packages @dzoccarato! You're right, for something like a basic 'stub' testing entity, it would probably make more sense to live in `AndcultureCode.CSharp.Testing`. I believe we have done a similar thing in `AndcultureCode.CSharp.Extensions` in order to test some extension methods there. At this point, we need something simple & reusable for multiple packages. The [`UserStub`](https://github.com/AndcultureCode/AndcultureCode.CSharp.Extensions/blob/main/test/AndcultureCode.CSharp.Extensions.Tests/Stubs/UserStub.cs) in that package has a few fields that provide flexibility. If you'd like, feel free to open up an issue in that repo and port over that entity + its [factory](https://github.com/AndcultureCode/AndcultureCode.CSharp.Extensions/blob/main/test/AndcultureCode.CSharp.Extensions.Tests/Factories/UserStubFactory.cs).

The extensions repository does have an example on how these factories are typically implemented & registered (in a base unit test class). I would not recommend in-line defining them as you have here - 

```CS
// RepositoryConductorTest.cs:761-770
FactoryExtensions.DefineFactory(new TestEntityFactory(), () => new TestEntity() { Id = 0, Name = "Alice" });
var query = BuildList<TestEntity>(10)
            .AsQueryable()
            .GroupBy(
                entity => entity.Name,
                (name, list) => new GroupingResult
                {
                    Key = name,
                    Value = list.Count()
                });
```

One thing to note about that `UserStub` is that it does not actually implement `IEntity`, which we would need for these repository tests. Be sure to make that change when porting.

wintondeshong reviewed (approved):
@dzoccarato thanks for making these changes 👍

wintondeshong commented:
@brandongregoryscott I actually need this change for a project I'm on. I've given this branch a run locally and verified it works as expected. If there are outstanding items you want changed, mind adding github issue? Gonna merge and ask for forgiveness later :P

dzoccarato commented:
> @dzoccarato thanks for making these changes 👍 - i really appreciate the level of effort and thoroughness you've taken to implement this across the various packages

@wintondeshong  thanks! :)

brandongregoryscott commented:
@all-contributors add @dzoccarato for code, test, doc
